Convert TF models using TF Python API.

dellis23 opened the pull request:
Not ready for review yet. I need to test the tool itself with the benchmarks.

dellis23 committed:
Convert TF models using TF Python API.

dellis23 committed:
Try installing tf nightly just for tests.

This should be reverted before merge, once the required
functionality makes its way into TF main.

dellis23 committed:
Replace cpp tool with python version.

dellis23 committed:
Update docker image to include tf-nightly.

Upgrade pip in docker image.

Commit forgotten change to tf integration.

stellaraccident commented on compiler/bindings/python/iree/compiler/tools/tf.py:
```diff
@@ -200,6 +217,13 @@ def compile_saved_model(saved_model_dir: str, **kwargs):
+    mlir = get_mlir(saved_model_dir, exported_names=options.exported_names)
+    temp_file.write(mlir)
+    temp_file.flush()
+    import_cl = ["cat", temp_file.name]
```
that's weird. I probably would have just used the contents directly vs redirecting through `cat`

dellis23 replied:
Oh weird, somehow I missed this comment.  Yeah, I considered reworking that too but was just wanting to get something working and was trying to re-used the existing command execution pipeline code.  Let me know if you want me to read and feed through stdin in Python instead.

stellaraccident replied:
Can be fixed post. I expect this will fail on Windows, but none of this has worked on Windows in a long time, afaik.

dellis23 committed:
Migrate contents of compiler tool to iree_import_tf script.

GMNGeoffrey commented on integrations/tensorflow/iree_tf_compiler/BUILD.bazel:
```diff
@@ -71,24 +71,10 @@ cc_binary(
 )
 
-cc_binary(
+py_binary(
```
Just delete this. There is no need to "build" a binary for Python. We have elected to completely bail on trying to use Bazel for Python dependency management. It may be desirable to move the tool into a separate directory so that it's clearly distinct from these compile C++ binaries (which should also go away soon). Then there are no symlinks: the CMake build just directly references the Python executable script.

dellis23 replied:
(I assume this is a slightly delayed comment, so resolving. I removed this in the previous commit.)

dellis23 committed:
Remove old python file.

dellis23 committed:
Add support for arguments the benchmark tool is calling with.

dellis23 committed:
Change first argument to positional argument.

dellis23 committed:
Remove iree-import-tf from importer-binaries filegroup.

dellis23 committed:
Attempt to fix pytype / lint issues.

The pytype issue wasn't caused here, but I'll try to fix it anyway.

See: https://github.com/openxla/iree/pull/12554#issuecomment-1460934174

dellis23 committed:
Don't attempt to copy iree-import-tf binary.

dellis23 committed:
Move TF import inside get_mlir function.

dellis23 committed:
Add absl dependency.

dellis23 committed:
Remove unnecessary explicit absl dependency, fix tf-nightly

dellis23 committed:
Remove command-line argument building based on previous tool.

dellis23 committed:
Fix missing options.

dellis23 committed:
Add tfs back in.

I guess we are using this even though the scope of the context manager
has exited...

dellis23 committed:
Remove some failing tests.

These I was able to reasonably grep / cut / rm.  The remainder
I'll have to find and remove manually.

dellis23 committed:
Remove remaining (hopefully) failed tests.

dellis23 committed:
Remove failing GPU test.

dellis23 committed:
DEBUG: see what's in binaries archive

dellis23 committed:
Attempt to install iree_tf

dellis23 committed:
Install python / pip.

dellis23 committed:
Remove check for symlink_binaries.sh-generated file.

dellis23 committed:
Don't look for iree-import-tf on binaries path.

dellis23 committed:
Try to find where it was installed.

dellis23 committed:
Check bin.

dellis23 committed:
Check more locations.

dellis23 committed:
Search everywhere.

dellis23 committed:
Just try runing with the full path.

dellis23 committed:
Install iree-import-tf inside docker container.

dellis23 committed:
Debug installation path inside docker container.

dellis23 committed:
Run from local bin.

dellis23 committed:
Install compiler.

dellis23 committed:
Submodule update.

dellis23 committed:
Install TF nightly.

dellis23 committed:
Disable TF benchmarks.

dellis23 committed:
Comment rest of file.

dellis23 committed:
Remove yaml comments to fix linter.

ScottTodd commented on .github/workflows/ci.yml:
```diff
@@ -673,6 +673,13 @@ jobs:
         run: gcloud storage cp "${TF_BINARIES_GCS_ARTIFACT}" "${TF_BINARIES_ARCHIVE}"
       - name: "Extracting TF binaries archive"
         run: tar -xf "${TF_BINARIES_ARCHIVE}"
+      - name: "DEBUG: see what's in the binaries archive"
+        run: find $(dirname ${TF_BINARIES_ARCHIVE}) -maxdepth 2
```
This is rough around the edges, but I was just able to reproduce what the CI does locally with this sequence of commands: https://gist.github.com/ScottTodd/bcb137207c7320865442f024e61c61d2

The trickiest details are in how Docker handles mounted directories and how artifacts are copied between jobs

dellis23 replied:
Re: TF_BINARIES_OUTPUT_DIR, what commit or branch did you run this on?  The new tool doesn't end up in TF_BINARIES_OUTPUT_DIR, since we're no longer building with bazel / copying around symlinks.  It's installed when the package is pip installed (although, as you can see from my commits, the environment still isn't aware of the installed commands).

ScottTodd replied:
I'm forked off of https://github.com/openxla/iree/pull/12758/commits/226434815b15bce74e0b4d58d518bca2e3827fe1.

Working here: https://github.com/ScottTodd/iree/tree/tf-python-api

For local iteration I've commented out most of `tests/e2e/test_artifacts/generated_e2e_test_iree_artifacts.cmake`, so I can just run the `build_e2e_test_artifacts.sh` step and see a single set of error messages from `iree_import_tf_model` (next I was going to try the pip install commands you added)

dellis23 replied:
I need to disable 5 tests in `generated_e2e_test_iree_artifacts.cmake`, but I assume it's not a good idea to modify that file directly.  I've tried to traverse the numerous layers involved but haven't figured out the core source of truth yet.  Going to continue on it tomorrow.

ScottTodd replied:
instructions: https://github.com/openxla/iree/tree/main/tests/e2e/test_artifacts#updating-test-artifacts (and maybe @pzread or someone else could help you with the specific targets)

dellis23 replied:
Is something going on with github CI right now? I can't get anything but the CLA agreement test to run.  Only 7 other jobs are showing up, and they all say `Expected — Waiting for status to be reported`.

GMNGeoffrey replied:
Should this step have been removed before merging?

dellis23 committed:
Fix import for e2e test artifacts.

dellis23 committed:
Attempt to remove e2e models.

dellis23 committed:
Attempt to retrigger CI.

dellis23 committed:
Retrigger CI.

dellis23 committed:
Regenerate cmake files after conflict fix.

dellis23 committed:
Resolve conflicts.

dellis23 committed:
Fix incorrect merge conflict.

Clean up some stuff and add note about disabling tests.

dellis23 committed:
Try to make TF note stand out a bit more.

dellis23 committed:
Fix spelling mistake.

jpienaar commented on integrations/tensorflow/test/requirements.txt:
```diff
@@ -1,15 +1,4 @@
 # Requirements for running TF tests
-tensorflow>=2.7.1
+tf-nightly
```
Should we pin to a specific one?

dellis23 replied:
I guess we could temporarily.  All we're doing really is waiting for the next TF release, at which point we should be able to get back on the main TF releases.

jpienaar replied:
Yes, my concern is just that the testing of nightly is lower, so I'd prefer not flakes from TF updates and I'm ok to pinning to any working one until next release.

dellis23 replied:
Makes sense, updated

dellis23 committed:
Pin to nightly release and add note about public release.

stellaraccident reviewed (approved):
I deem the drop in test coverage a necessary tradeoff for moving forward. In practice, the TFLite and Jax test suites have long since been the high ground here driving current work. I will note that, in theory, most of this test coverage could be retained with a re-engineering of the TF regression test tooling to make it not assume a 1:1 correspondence between the TF-presented API and the IREE-exposed API. This would make it fairly similar to the level of support that tf-compile has (as an example), but with state/variables modeled.

The will to keep this programming model alive has long since lapsed, and we are better spending our time on current generation testing efforts and framework features.

dellis23 committed:
Resolve newly-pushed conflicts.

GMNGeoffrey reviewed (commented):
Yeah I'm concerned that this dropped a whole bunch of our benchmarks, many of which were just added recently (not to mention tests). If the plan is to fix forward, who owns doing that? Who owns chasing down the performance regressions we will miss until that completes?
